packages/vscode-extension/src/providers/MaestroCodeLensProvider.ts

```diff
@@ -0,0 +1,58 @@
+import {
+  CodeLensProvider,
+  TextDocument,
+  CancellationToken,
+  CodeLens,
+  Range,
+  Command,
+} from "vscode";
+
+export class MaestroCodeLensProvider implements CodeLensProvider {
+  provideCodeLenses(
+    document: TextDocument,
+    token: CancellationToken
+  ): CodeLens[] | Thenable<CodeLens[]> {
+    if (!this.checkMaestroFile(document)) {
+      return [];
+    }
+
+    const command: Command = {
+      title: "Run Maestro test",
+      command: "RNIDE.startMaestroTest",
+      arguments: [[document.fileName]],
+    };
+
+    const codeLenses: CodeLens[] = [];
+    codeLenses.push(new CodeLens(this.createRange(document, 0), command));
+
+    return codeLenses;
+  }
+
+  checkMaestroFile(document: TextDocument): boolean {
+    // To fairly certainly identify a Maestro test file that we can run, it must:
+    // not be named config.yaml or .yml
+    // include "appId" (url on web, but we only do mobile)
+    // include "---" splitting configuration and steps
+    // include valid steps beginning with "-" that are not comments
+    const text = document.getText();
+    if (
+      document.fileName.endsWith("config.yaml") ||
+      document.fileName.endsWith("config.yml") ||
+      !text.includes("appId:")
+    ) {
+      return false;
+    }
+    const splitText = text.split("---");
```

Comment:
You could probably look for a _line that starts with_ the three dashes, something like
```suggestion
    const splitText = text.split(/^\s*---/m);
```
to avoid matching on files that have the dashes somewhere in the text.